design/tla/Task.tla

```diff
@@ -0,0 +1,760 @@
+  \/ Agent
+  \/ Reaper
+  \* For model checking: don't report deadlock if we're limiting events
+  \/ (nEvents = maxEvents /\ UNCHANGED vars)
+
+(* This is a ``temporal formula''. It takes a sequence of states representing the
+   changing state of the world and evaluates to TRUE if that sequences of states is
+   a possible behaviour of SwarmKit. *)
+Spec ==
+  \* The first state in the behaviour must satisfy Init:
+  /\ Init
+  \* All consecutive pairs of states must satisfy Next or leave `vars' unchanged:
+  /\ [][Next]_vars
+  (* Some actions are required to happen eventually. For example, a behaviour in
+     which SwarmKit stops doing anything forever, even though it could advance some task
+     from the `new' state, isn't a valid behaviour of the system.
+     This property is called ``weak fairness''. *)
+  /\ WF_vars(OrchestratorProgress \/ Allocator \/ Scheduler \/ AgentProgress \/ Reaper)
+     (* We don't require fairness of:
+        - User (we don't control them),
+        - RestartTask (services aren't required to be updated),
+        - RejectTask (tasks aren't required to be rejected),
+        - ContainerExit (we don't specify image behaviour) or
+        - WorkerDown (workers aren't required to fail). *)
+
+-------------------------------------------------------------------------------
+\* Properties to verify
+
+(* These are properties that should follow automatically if the system behaves as
+   described by `Spec' in the previous section. *)
+
+\* A state invariant (things that should be true in every state).
+Inv ==
+  \A id \in DOMAIN tasks :
+    (* Every task has a service:
+
+       XXX: The spec says: ``In some cases, there are tasks that exist independent of any service.
+            These do not have a value set in service_id.''. Need an example of one. *)
+    /\ tasks[id].service \in DOMAIN services
+    \* Tasks have nodes once they reach `assigned':
+    /\ assigned \preceq State(id) => tasks[id].node \in Node
+
+\* A special ``state'' used when a task doesn't exist.
+null == "null"
+
+\* All the possible transitions
+Transition == {
+  << null, new >>,
+  << new, pending >>,
+  << pending, assigned >>,
+
+  << assigned, accepted >>,
+  << accepted, preparing >>,
+  << preparing, ready >>,
+  << ready, starting >>,
+  << starting, running >>,
+
+  << assigned, rejected >>,
+  << accepted, rejected >>,
+  << preparing, rejected >>,
+  << ready, rejected >>,
+  << starting, rejected >>,
+
+  << running, complete >>,
+  << running, failed >>,
+
+  << running, shutdown >>,
+
+  << shutdown, remove >>,
+  << rejected, remove >>,
+  << complete, remove >>,
+  << failed, remove >>,
+
+  << assigned, orphaned >>,
+  << accepted, orphaned >>,
+  << preparing, orphaned >>,
+  << ready, orphaned >>,
+  << starting, orphaned >>,
+  << rejected, orphaned >>,
+  << running, orphaned >>,
+  << complete, orphaned >>,
+  << failed, orphaned >>,
+  << shutdown, orphaned >>,
+
+  << remove, null >>,
+  << orphaned, null >>
```

Comment:
Tasks can be deleted (moved to the null state, I think, in terms of the implementation of this model) from any terminal state. For example, when a task history is deleted, the old tasks will all be in terminal states (like `SHUTDOWN` or `FAILED`) and will be deleted straight from those states. Once a task is in a terminal state, the resources it had in use are definitely free and it is safe to remove it.